File: Assignment_A2/Pass1.java
package Assignment_A2;

import java.util.*;
import java.io.*;

public class Pass1 implements Serializable{
    class MNTableEntry implements Serializable{
        public String macroName;
        public int numKPD = 0;
        public int numPP = 0;
        public int MDTabPtr = 0;
        public int kpdTabPtr = 0;

        public MNTableEntry(){}

        public MNTableEntry(String name,int k,int p,int mPtr,int kPtr){
            this.macroName = name;
            this.numKPD = k;
            this.numPP = p;
            this.MDTabPtr = mPtr;
            this.kpdTabPtr = kPtr;
        }
    }

    class MDTableEntry implements Serializable{
        public String mnemonic;
        public String operand1;
        public String operand2;
        public int op1Index = -1;
        public int op2Index = -1;
    }

    private ArrayList<String[]> tokens = new ArrayList<>() ;
    private ArrayList<String[]> KDPTAB = new ArrayList<>() ;
    private ArrayList<String> PNTAB = new ArrayList<>() ;  
    private ArrayList<MNTableEntry> MNT = new ArrayList<>() ; 
    private ArrayList<MDTableEntry> MDT = new ArrayList<>() ; 
    private String currentMacroName ; 
    private HashMap<String , List<String>> pnTabHashMap = new HashMap<>();

    public Pass1( String sourceFilepath ) {
        ArrayList<String> lines = readLines( sourceFilepath ) ; 
        for( String line : lines ) {
            tokens.add( line.split(" ") ) ; 
        }
    }

    public void perform() {
        int mdtabPtr = 0 ;
        int kpdtabPtr = 0 ;
        for( int i = 1 ; i < tokens.size() ; i++ ) { 
            String[] lineTokens = tokens.get( i ) ; 
            System.out.println( Arrays.toString( lineTokens ) ) ;
            if( tokens.get( i - 1 )[0].equals( "MACRO" ) )  {
                currentMacroName = lineTokens[ 0 ] ; 
                int numKPD = 0 ; 
                int numPP = 0 ; 
                PNTAB.clear();
                for( int j = 1 ; j < lineTokens.length ; j++ ) {
                    String parameter = lineTokens[ j ] ; 
                    if( parameter.contains( "=" ) ) {
                        KDPTAB.add( parameter.split( "=" ) ); 
                        numKPD++ ; 
                    }
                    else {
                        numPP++ ; 
                    }
                    System.out.println( "Parameter added : " + parameter ) ; 
                    PNTAB.add( parameter.split("=")[0] ) ; 
                }
                MNT.add( new MNTableEntry( currentMacroName , numKPD , numPP , mdtabPtr , kpdtabPtr ) ) ;
                System.out.println( "Added to map : " + currentMacroName ) ; 
                kpdtabPtr += numKPD ; 
                pnTabHashMap.put( currentMacroName , PNTAB ) ;  
            }
            else if( !lineTokens[0].equals( "MACRO" ) &&
             !lineTokens[0].equals( "MEND" ) ) {
                MDTableEntry entry = new MDTableEntry(); 
                entry.mnemonic = lineTokens[0] ;   
                List<String> currentPNTAB = pnTabHashMap.get( currentMacroName ) ; 
                if( lineTokens[1].startsWith( "&" ) ) {
                    int index = currentPNTAB.indexOf(lineTokens[1]) ; 
                    entry.operand1 = lineTokens[1] ;
                    if( index >= 0 ) { 
                        entry.op1Index = index ; 
                    }
                }
                if( lineTokens[2].startsWith( "&" ) ) {
                    int index = currentPNTAB.indexOf(lineTokens[2]) ; 
                    entry.operand2 = lineTokens[2] ;
                    if( index >= 0 ) { 
                        entry.op2Index = index ; 
                    }
                }
                MDT.add( entry ) ; 
            }
            else if( lineTokens[0].equals( "MEND" ) ) {
                MDTableEntry entry = new MDTableEntry(); 
                entry.mnemonic = "MEND" ;
                MDT.add(entry) ;    
            }
            mdtabPtr++ ; 
        }

        saveTable( MNT , "Assignment_A2/MNT.dat" ) ;
        saveTable( KDPTAB , "Assignment_A2/KPDTAB.dat" ) ;
        saveTable( pnTabHashMap , "Assignment_A2/PNTAB.dat" ) ;
        saveTable( MDT , "Assignment_A2/MDT.dat" ) ;

        printPNTAB();
        printMNTAB();
        printKPDTAB();
        
    }

    private static void saveTable( Object table , String filepath ) {
        try {
            FileOutputStream fos = new FileOutputStream( new File( filepath ) ) ; 
            ObjectOutputStream outputStream = new ObjectOutputStream(fos) ; 
            outputStream.writeObject( table );
            outputStream.flush();
            outputStream.close();
            fos.close();
        }
        catch( FileNotFoundException e ){
            e.printStackTrace();
        }
        catch( IOException e ) {
            e.printStackTrace(); 
        }
    }

    private static ArrayList<String> readLines( String filePath ) {
        File textFile = new File( filePath ) ;
        ArrayList<String> lines = new ArrayList<>();  
        try {
            FileInputStream inputStream = new FileInputStream( textFile ) ; 
            BufferedReader reader = new BufferedReader( new InputStreamReader( inputStream ) ) ; 
            while( reader.ready() ) {
                lines.add( reader.readLine() ) ; 
            }
            inputStream.close();
        }
        catch( FileNotFoundException e ) {
            e.printStackTrace();
        }
        catch( IOException e ) {
            e.printStackTrace();
        }
        return lines ; 
    }

    private void printMNTAB() {
        for( MNTableEntry entry : MNT ) {
            System.out.println( 
                entry.macroName + " " + 
                entry.numKPD + " " + 
                entry.numPP + " " + 
                entry.MDTabPtr + " " + 
                entry.kpdTabPtr );
        }
    }

    private void printKPDTAB() {
        for( String[] entry : KDPTAB ) {
            System.out.println( entry[0] + " " + entry[1] );
        }
    }

    private void printPNTAB() {
        Iterator<String> it = pnTabHashMap.keySet().iterator() ; 
        while( it.hasNext() ) {
            String macroName = it.next() ; 
            System.out.println( "Macro Name: " + macroName ) ; 
            List<String> params = pnTabHashMap.get( macroName ) ; 
            for( String param : params ) {
                System.out.println( param ) ; 
            }
        }
    }
}

File: Assignment_A2/Python/pass1.py
class MNTableEntry:
    def __init__(self, name) -> None:
        self.name = name
        self.num_kpd = 0
        self.num_pp = 0
        self.mdtab_ptr = 0
        self.kpdtab_ptr = 0


class MDTableEntry:
    def __init__(self, mnemonic_name) -> None:
        self.mnemonic = mnemonic_name
        self.operand1 = ""
        self.operand2 = ""
        self.op1_index = -1
        self.op2_index = -1


mntab: list[MNTableEntry] = []
mdtab: list[MDTableEntry] = []
kpdtab: list[tuple[str, str]] = []
list_of_pntabs: dict[str, list[str]] = {}

source_path = input("Enter file path: ")
with open(source_path, 'r') as file:
    source_lines = file.read().split('\n')
    source_tokens = [line.split() for line in source_lines]

mdtab_pointer = 0
kpdtab_pointer = 0
name = ""

for i in range(1, len(source_tokens)):
    line_tokens = source_tokens[i]

    if source_tokens[i-1][0] == 'MACRO':
        name = line_tokens[0]
        parameters = line_tokens[1:]
        pntab: list[str] = []
        kpd_counter = 0
        pp_counter = 0
        p_name = ''

        for parameter in parameters:
            if '=' in parameter:
                p_name, val = parameter.split('=')
                kpdtab.append((p_name, val))
                kpd_counter += 1
            else:
                p_name = parameter
                pp_counter += 1
            pntab.append(p_name)

        mntab_entry = MNTableEntry(name=name)
        mntab_entry.num_kpd = kpd_counter
        mntab_entry.num_pp = pp_counter
        mntab_entry.kpdtab_ptr = kpdtab_pointer
        mntab_entry.mdtab_ptr = mdtab_pointer
        mntab.append(mntab_entry)

        list_of_pntabs[name] = pntab
        kpdtab_pointer += kpd_counter

    elif line_tokens[0] != 'MEND' and line_tokens[0] != 'MACRO':
        mdtab_entry = MDTableEntry(mnemonic_name=line_tokens[0])
        pntab: list[str] = list_of_pntabs[name]

        parameter = line_tokens[1]
        if parameter in pntab:
            mdtab_entry.op1_index = pntab.index(parameter)
        mdtab_entry.operand1 = parameter

        parameter = line_tokens[2]
        if parameter in pntab:
            mdtab_entry.op2_index = pntab.index(parameter)
        mdtab_entry.operand2 = parameter

        mdtab.append(mdtab_entry)
        mdtab_pointer += 1

    elif line_tokens[0] == 'MEND':
        mdtab_entry = MDTableEntry('MEND')
        mdtab_pointer += 1
        mdtab.append(mdtab_entry)


print( "------------ MNTAB ----------------")   
print( "Name".ljust( 10 ) , "#PP".ljust( 5 ) , "#KP".ljust( 5 ) , "KPDTAB_PTR".ljust( 10 ) , "MDTAB_PTR".ljust( 10 ) )  
for entry in mntab:
    print( 
        entry.name.ljust( 10 ) , 
        str(entry.num_pp).ljust( 5 ) , 
        str(entry.num_kpd).ljust( 5 ) , 
        str(entry.kpdtab_ptr).ljust( 10 ) , 
        str(entry.mdtab_ptr).ljust( 10 )
    )

print( "------------ MDTAB ----------------")   
for entry in mdtab:
    print( 
        entry.mnemonic.ljust( 10 ) , 
        (entry.operand1 if entry.op1_index == -1 else f"(P,{entry.op1_index})").ljust(10) , 
        (entry.operand2 if entry.op2_index == -1 else f"(P,{entry.op2_index})").ljust(10) ,  
    )
    
print( "------------ KPDTAB ----------------")   
print( "Name".ljust( 10 ) , "Default Value".ljust( 15 ) )
for param in kpdtab:
    print( param[0].ljust( 10 ) , param[1].ljust( 15 ) )

print( "------------- PNTAB -----------------")
for ( name , pntab ) in list_of_pntabs.items():
    print( "PNTAB for macro" , name ) 
    for param in pntab:
        print( param )